[ResourceMonitor] Track network usage of ServiceWorker.

basuke opened the pull request:
#### ab04d94a385b7bf463155ce2fc89193bd317310f
<pre>
[ResourceMonitor] Track network usage of ServiceWorker.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287506">https://bugs.webkit.org/show_bug.cgi?id=287506</a>
<a href="https://rdar.apple.com/144586530">rdar://144586530</a>

Reviewed by Youenn Fablet.

We don&apos;t track network usage of ServiceWorker registered by iframe and also there&apos;s no direct connection
between ServiceWorker and iframe right now. To monitor network usage, send IPC message to report
network usage of ServiceWorker to all connected iframe document.

To make this happen:
1. Add new parameter, `workerIdentifier` in NetworkResourceLoadParameters.
2. Report network usage of the ServiceWorker from NetworkResourceLoader to connected ServiceWorker clients.
3. Tell ResourceMonitor to treate those usage as part of the resource usage of iframe.

* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/iframe2.html:
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/service-worker-setup.js: Added.
(async waitForState):
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/service-worker.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/service-worker.js: Added.
(async e):
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/unregister-service-worker.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/using-1k.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/using-6k.html: Added.
* LayoutTests/http/tests/iframe-monitor/shared-worker.html: Removed.
* LayoutTests/http/tests/iframe-monitor/workers/service-worker-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/workers/service-worker-not-count-twice-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/workers/service-worker-not-count-twice.html: Added.
* LayoutTests/http/tests/iframe-monitor/workers/service-worker.html: Added.
* LayoutTests/http/tests/iframe-monitor/workers/shared-worker-expected.txt: Renamed from LayoutTests/http/tests/iframe-monitor/shared-worker-expected.txt.
* LayoutTests/http/tests/iframe-monitor/workers/shared-worker.html: Added.
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/ResourceLoaderOptions.h:
* Source/WebCore/loader/WorkerThreadableLoader.cpp:
(WebCore::m_contextIdentifier):
* Source/WebCore/workers/FetchingWorkerIdentifier.h: Added.
* Source/WebCore/workers/service/SWClientConnection.cpp:
(WebCore::SWClientConnection::reportNetworkUsageToWorkerClient):
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::reportNetworkUsageToAllWorkerClients):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebCore/workers/shared/SharedWorkerObjectConnection.cpp:
(WebCore::SharedWorkerObjectConnection::reportNetworkUsageToWorkerObject):
* Source/WebCore/workers/shared/SharedWorkerObjectConnection.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
(WebKit::NetworkResourceLoadParameters::NetworkResourceLoadParameters):
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.serialization.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::didFinishLoading):
(WebKit::NetworkResourceLoader::sendDidReceiveDataMessage):
(WebKit::NetworkResourceLoader::updateBytesTransferredOverNetwork):
(WebKit::NetworkResourceLoader::reportNetworkUsageToAllSharedWorkerObjects):
(WebKit::NetworkResourceLoader::reportNetworkUsageToAllServiceWorkerClients):
(WebKit::NetworkResourceLoader::reportNetworkUsageToAllSharedWorkers): Deleted.
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::resolveUnregistrationJobInClient):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::reportNetworkUsageToWorkerClient):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
(WebKit::WebSharedWorkerServer::reportNetworkUsageToAllSharedWorkerObjects):
(WebKit::WebSharedWorkerServer::reportNetworkUsageToAllSharedWorkerClients): Deleted.
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerConnection.cpp:
(WebKit::WebSharedWorkerServerConnection::reportNetworkUsageToWorkerObject):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.messages.in:
* Source/WebKit/WebProcess/Storage/WebSharedWorkerObjectConnection.messages.in:

Canonical link: <a href="https://commits.webkit.org/291262@main">https://commits.webkit.org/291262@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c42cbda35f120dff0c4b77f24a0eaabeed93cc7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92473 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/12017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1619 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97457 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/42980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94523 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12292 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20476 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70858 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/42980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95475 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/9340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/83748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/51189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/9036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/1385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99483 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19524 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/19774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/79628 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79144 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19610 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/19508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/19195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/22655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20935 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on LayoutTests/http/tests/iframe-monitor/resources/--eligible--/iframe2.html:
```diff
@@ -8,4 +8,4 @@
 
     const size = 1 * 1024;
     fetch(`../generate-byte.py?size=${size}`);
```
Why --eligible-- as folder name?

youennf commented on LayoutTests/http/tests/iframe-monitor/resources/--eligible--/service-worker-setup.js:
```diff
@@ -0,0 +1,39 @@
+async function setupServiceWorker(url, receiver) {
+    if (!("serviceWorker" in navigator)) {
```
We do not generally test this as service worker is always supported in WK2. We skip tests for WK1.

youennf commented on LayoutTests/http/tests/iframe-monitor/resources/--eligible--/service-worker.html:
```diff
@@ -0,0 +1,40 @@
+                window.parent.postMessage(true, '*');
+            } else {
+                window.parent.postMessage(false, '*');
+            }
```
window.parent.postMessage(data instanceof Blob, '*');

youennf commented on Source/WebCore/loader/WorkerThreadableLoader.cpp:
```diff
@@ -157,10 +158,10 @@ WorkerThreadableLoader::MainThreadBridge::MainThreadBridge(ThreadableLoaderClien
         InspectorInstrumentation::willSendRequest(*serviceWorkerGlobalScope, m_workerRequestIdentifier, request);
-
-    if (RefPtr sharedWorkerGlobalScope = dynamicDowncast<SharedWorkerGlobalScope>(globalScope))
+        optionsCopy->options.workerIdentifier = serviceWorkerGlobalScope->thread().identifier();
```
In case of service worker doing a fetch from a fetch event, aren't we adding twice the networking cost to the iframe that triggered the load?

basuke replied:
Sorry I don't get what case you mean. Can you show me concrete example?

youennf replied:
In case the iframe is doing fetch(url1), it would trigger a fetch event in service worker.
If the service worker is doing: `event.respondWith(fetch(event.request)))`, will we count the network cost of fetching url1 twice for the iframe?
Maybe this is fine like it is, but it is good we are sure how we count things.

youennf replied:
Ah actually, we do not count the network cost twice, please disregard the comment.

youennf commented on Source/WebCore/workers/service/server/SWServer.cpp:
```diff
@@ -1942,4 +1942,19 @@ void SWServer::Connection::retrieveRecordResponseBody(BackgroundFetchRecordIdent
+{
+    if (RefPtr worker = workerByID(identifier)) {
+        if (RefPtr connection = contextConnectionForRegistrableDomain(worker->topRegistrableDomain())) {
+            ServiceWorkerClientQueryOptions options = { true, ServiceWorkerClientType::All };
```
```suggestion
            ServiceWorkerClientQueryOptions options = { true, ServiceWorkerClientType::Window };
```
Since you only report for iframes.

youennf commented on Source/WebCore/workers/service/server/SWServer.cpp:
```diff
@@ -1942,4 +1942,19 @@ void SWServer::Connection::retrieveRecordResponseBody(BackgroundFetchRecordIdent
+    if (RefPtr worker = workerByID(identifier)) {
+        if (RefPtr connection = contextConnectionForRegistrableDomain(worker->topRegistrableDomain())) {
+            ServiceWorkerClientQueryOptions options = { true, ServiceWorkerClientType::All };
+            matchAll(*worker, options, [connection = WTFMove(connection), delta](auto&& clientDataList) {
```
```suggestion
            matchAll(*worker, options, [connection = connection.releaseNonNull(), delta](auto&& clientDataList) {
```

youennf commented on Source/WebCore/workers/service/server/SWServer.cpp:
```diff
@@ -1942,4 +1942,19 @@ void SWServer::Connection::retrieveRecordResponseBody(BackgroundFetchRecordIdent
 }
 
+#if ENABLE(CONTENT_EXTENSIONS)
+void SWServer::reportNetworkUsageToAllWorkerClients(ServiceWorkerIdentifier identifier, size_t delta)
```
```suggestion
void SWServer::reportNetworkUsageToAllWorkerClients(ServiceWorkerIdentifier identifier, size_t bytesTransferredOverNetworkDelta)
```
to align with header?

youennf commented on Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
```diff
@@ -65,7 +65,7 @@ NetworkResourceLoadParameters::NetworkResourceLoadParameters(
     , OptionSet<WebCore::HTTPHeadersToKeepFromCleaning> httpHeadersToKeep
     , std::optional<WebCore::FetchIdentifier> navigationPreloadIdentifier
-    , std::optional<WebCore::SharedWorkerIdentifier> workerIdentifier
+    , std::variant<std::monostate, WebCore::SharedWorkerIdentifier, WebCore::ServiceWorkerIdentifier> workerIdentifier
```
We could add a type like
using WorkerIdentifier = ...
to shorten the code.

basuke replied:
This is required to distinguish ServiceWorker and SharedWorker.

youennf replied:
I was meaning something like:
`using FetchingWorkerIdentifier =  std::variant<std::monostate, WebCore::SharedWorkerIdentifier, WebCore::ServiceWorkerIdentifier>`.
And then using FetchingWorkerIdentifier everywhere to ease readability.

youennf commented on Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
```diff
@@ -316,17 +316,19 @@ void WebSharedWorkerServer::terminateContextConnectionWhenPossible(const WebCore
     sharedWorker->forEachSharedWorkerObject([&](auto sharedWorkerObjectIdentifier, auto&) {
         if (auto* serverConnection = m_connections.get(sharedWorkerObjectIdentifier.processIdentifier()))
-            serverConnection->reportNetworkUsageToWorkerObject(sharedWorkerObjectIdentifier, bytesTransferredOverNetwork);
+        serverConnection->reportNetworkUsageToWorkerObject(sharedWorkerObjectIdentifier, bytesTransferredOverNetworkDelta);
```
```suggestion
            serverConnection->reportNetworkUsageToWorkerObject(sharedWorkerObjectIdentifier, bytesTransferredOverNetworkDelta);
```

youennf commented on Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
```diff
@@ -316,17 +316,19 @@ void WebSharedWorkerServer::terminateContextConnectionWhenPossible(const WebCore
+    RELEASE_LOG(SharedWorker, "WebSharedWorkerServer::reportNetworkUsageToAllSharedWorkerObjects: sharedWorkerIdentifier=%" PRIu64 ", sharedWorker=%p, bytesTransferredOverNetworkDelta=%zu", sharedWorkerIdentifier.toUInt64(), sharedWorker, bytesTransferredOverNetworkDelta);
     if (!sharedWorker)
-        return;
+    return;
```
unneeded change

youennf reviewed (commented):
Patch looks mostly ready to me. A question and some nits.
I am unsure whether the heuristic is exactly what we want, if one iframe is abusing, won't it report the abuse to all iframes (seems okay given they are all same origin, but I just want to check).
Also, should we penalize all iframes for shared worker traffic as well?

youennf reviewed (approved):
OK, lgtm, let's discuss as a follow-up whether the current shared worker reporting approach is what we want

basuke committed:
[ResourceMonitor] Track network usage of ServiceWorker.
https://bugs.webkit.org/show_bug.cgi?id=287506
rdar://144586530

Reviewed by Youenn Fablet.

We don't track network usage of ServiceWorker registered by iframe and also there's no direct connection
between ServiceWorker and iframe right now. To monitor network usage, send IPC message to report
network usage of ServiceWorker to all connected iframe document.

To make this happen:
1. Add new parameter, `workerIdentifier` in NetworkResourceLoadParameters.
2. Report network usage of the ServiceWorker from NetworkResourceLoader to connected ServiceWorker clients.
3. Tell ResourceMonitor to treate those usage as part of the resource usage of iframe.

* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/iframe2.html:
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/service-worker-setup.js: Added.
(async waitForState):
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/service-worker.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/service-worker.js: Added.
(async e):
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/unregister-service-worker.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/using-1k.html: Added.
* LayoutTests/http/tests/iframe-monitor/resources/--eligible--/using-6k.html: Added.
* LayoutTests/http/tests/iframe-monitor/shared-worker.html: Removed.
* LayoutTests/http/tests/iframe-monitor/workers/service-worker-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/workers/service-worker-not-count-twice-expected.txt: Added.
* LayoutTests/http/tests/iframe-monitor/workers/service-worker-not-count-twice.html: Added.
* LayoutTests/http/tests/iframe-monitor/workers/service-worker.html: Added.
* LayoutTests/http/tests/iframe-monitor/workers/shared-worker-expected.txt: Renamed from LayoutTests/http/tests/iframe-monitor/shared-worker-expected.txt.
* LayoutTests/http/tests/iframe-monitor/workers/shared-worker.html: Added.
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/ResourceLoaderOptions.h:
* Source/WebCore/loader/WorkerThreadableLoader.cpp:
(WebCore::m_contextIdentifier):
* Source/WebCore/workers/FetchingWorkerIdentifier.h: Added.
* Source/WebCore/workers/service/SWClientConnection.cpp:
(WebCore::SWClientConnection::reportNetworkUsageToWorkerClient):
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::reportNetworkUsageToAllWorkerClients):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebCore/workers/service/server/SWServerToContextConnection.h:
* Source/WebCore/workers/shared/SharedWorkerObjectConnection.cpp:
(WebCore::SharedWorkerObjectConnection::reportNetworkUsageToWorkerObject):
* Source/WebCore/workers/shared/SharedWorkerObjectConnection.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.cpp:
(WebKit::NetworkResourceLoadParameters::NetworkResourceLoadParameters):
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.h:
* Source/WebKit/NetworkProcess/NetworkResourceLoadParameters.serialization.in:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::didFinishLoading):
(WebKit::NetworkResourceLoader::sendDidReceiveDataMessage):
(WebKit::NetworkResourceLoader::updateBytesTransferredOverNetwork):
(WebKit::NetworkResourceLoader::reportNetworkUsageToAllSharedWorkerObjects):
(WebKit::NetworkResourceLoader::reportNetworkUsageToAllServiceWorkerClients):
(WebKit::NetworkResourceLoader::reportNetworkUsageToAllSharedWorkers): Deleted.
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::resolveUnregistrationJobInClient):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp:
(WebKit::WebSWServerToContextConnection::reportNetworkUsageToWorkerClient):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.cpp:
(WebKit::WebSharedWorkerServer::reportNetworkUsageToAllSharedWorkerObjects):
(WebKit::WebSharedWorkerServer::reportNetworkUsageToAllSharedWorkerClients): Deleted.
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServer.h:
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerConnection.cpp:
(WebKit::WebSharedWorkerServerConnection::reportNetworkUsageToWorkerObject):
* Source/WebKit/NetworkProcess/SharedWorker/WebSharedWorkerServerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.messages.in:
* Source/WebKit/WebProcess/Storage/WebSharedWorkerObjectConnection.messages.in:

Canonical link: https://commits.webkit.org/291262@main
